core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -311,6 +313,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
+        transportService.setLocalNode(localNode);
+        clusterService.add(transportService.getTaskManager());
+
+        // state NodeConnectionService last, so it's listeners will be first and last in cluster service
```

Comment:
> that sucks, sorry :) 

I agree! we should review the listeners and solve this. I review the addFirst/addLast usage  while some of them can be removed (it's either not important or they should use an observer - like in the restore service case), some are still order dependent. I personally would like it if the order was super clear and in one place, so if you change something you immediately see the implications. This is one of the things that long bugged me in the update task priority (though it's much more difficult to change there). 

Concretely I would propose doing the following:
1) I changed this PR to explicitly use NodeConnectionsService - I don't want to tackle the listeners infra in one go - this would make the change too big.
2) Open a follow up issue and discuss how to changes the "fixed" listeners (as I said, I'm in favor of upfront, before/during start and centralized registration). I will work on that or find someone who wants to.